### ISSUE TITLE
[Feature Request] Consider switching to CodeMirror for DevTools

### ISSUE THREAD

**Issue by Yash-Singh1** [darkreader/darkreader#4108]
**Describe the solution you'd like**

Currently the stylesheet editor and the DarkReader DevTools use `textarea` to render themselves. It does not have any syntax highlighting, current line highlighting, line numbers, etc. I think that CodeMirror could provide more features for the editors.

**Comment by Gusted** [comment]
Hi @Yash-Singh1!

>  It does not have any syntax highlighting, current line highlighting, line numbers

Is this a huge problem? The devtools is meant for simple CSS Fixes and should be as simple as it is.

Some 'concerns'.
1. CodeMirror will increase Dark Reader's binary with 300-400kB for just the source itself.
2. CodeMirror's provides a lot of settings where users probably wants to change them, this will require a new UI design in it's current window which is already small.
3. CodeMirror won't work without any modifcations(to the source code) to work with the dev tools. e.g.
```CSS
academic.microsoft.com

INVERT
.loader

CSS
.legend > rect:nth-child(1) {
    fill: unset !important;
}

================================
```
You can see their are KeyWords `CSS/INVERT` and the site which will be wrongly syntaxed highlight and a line seperator `====` which will throw errors in the CSS Parser/Linter of CodeMirror.
4. I *personally*  think the devtools should stay simplified.

@alexanderby Do you have any thoughts on this?

Regards,
Gusted

**Comment by Yash-Singh1** [comment]
Hey @Gusted!

> CodeMirror will increase Dark Reader's binary with 300-400kB for just the source itself.

This is not true. 99% of CodeMirror's repository is filled with addons and other things like that. There are just a few scripts and stylesheets which we really need. Also, we can just import these things using `unpkg`:
https://unpkg.com/

> CodeMirror's provides a lot of settings where users probably wants to change them, this will require a new UI design in it's current window which is already small.

This is not really required. We don't need to provide users with features such as theme customization, vim mode, etc.

> CodeMirror won't work without any modifcations(to the source code) to work with the dev tools. e.g.

We can add in our own mode.

All we have to do to add in CodeMirror is the following:
```js
CodeMirror.fromTextArea(document.getElementById("textArea"), {
  theme: "cobalt",
  mode: "myMODE",
  lineNumbers: true
});
```
Of course we need to import a few stylesheets and the scripts before we can do this. I think the main part of adding in CodeMirror will be adding in our own mode. Everything else can take up to just 10 lines of code.

**Comment by Gusted** [comment]
> This is not true. 99% of CodeMirror's repository is filled with addons and other things like that. There are just a few scripts and stylesheets which we really need. Also, we can just import these things using `unpkg`:
> https://unpkg.com/

Afaik what I saw the standalone version was already 380kb and the minified version 175Kb. Which has to be imported into the final binary which will take up the approx space. Also using a third party to fetch the library is a big no go for dark readers code base.




> This is not really required. We don't need to provide users with features such as theme customization, vim mode, etc.

Users will disagree or agree on that as everyone have their own quirk =). 



> We can add in our own mode.
> 
> All we have to do to add in CodeMirror is the following:
> 
> ```js
> CodeMirror.fromTextArea(document.getElementById("textArea"), {
>   theme: "cobalt",
>   mode: "myMODE",
>   lineNumbers: true
> });
> ```
> 
> Of course we need to import a few stylesheets and the scripts before we can do this. I think the main part of adding in CodeMirror will be adding in our own mode. Everything else can take up to just 10 lines of code.

Hmm, I will fiddle around the next few days with codeMirror and customMode.

**Comment by Yash-Singh1** [comment]
> Afaik what I saw the standalone version was already 380kb and the minified version 175Kb. Which has to be imported into the final binary which will take up the approx space. Also using a third party to fetch the library is a big no go for dark readers code base.

Yes, but we won't import everything. We will just import one or two scripts and stylesheets.

**Comment by Gusted** [comment]
> Yes, but we won't import everything. We will just import one or two scripts and stylesheets.

Afaik this still requires the other sourcecode and core function to make this work.

**Comment by Yash-Singh1** [comment]
I am working on a project using CodeMirror and here is what I had to import:
```html
    <script src="https://unpkg.com/codemirror@5.58.2/lib/codemirror.js"></script>
    <link rel="stylesheet" href="https://unpkg.com/codemirror@5.58.2/lib/codemirror.css" />
    <link rel="stylesheet" href="https://unpkg.com/codemirror@5.58.2/theme/cobalt.css" />
    <script src="https://unpkg.com/codemirror@5.58.2/mode/css/css.js"></script>
```
The `codemirror.js` file is the main source code allowing you to run CodeMirror inside the site (or extension).
The `codemirror.css` does some basic css for the editor.
The `cobalt.css` file is the theme that I imported.
The `css.js` parses and initializes the css mode.

There are also some other small addons which I don't think is necessary. My app is now functioning:

![Screenshot from 2020-11-14 17-14-59](https://user-images.githubusercontent.com/53054099/99160509-f5787d00-269c-11eb-877a-1a1214aa6bc8.png)

**Comment by Gusted** [comment]
> If I have some code like this:
> 
> ```
> academic.microsoft.com
> 
> INVERT
> .loader
> 
> CSS
> .legend > rect:nth-child(1) {
>     fill: unset !important;
> }
> 
> ================================
> ```
> 
> You can tell CodeMirror to syntax highlight the:
> 
> ```css
> .legend > rect:nth-child(1) {
>     fill: unset !important;
> }
> ```
> 
> And leave the rest to ourselves.

How does the `site` looks like? is it being highlighted? And the `INVERT` and `CSS` and the `===` line separator? those are my concerns.

**Comment by Yash-Singh1** [comment]
I was just pointing out that we don't need to be concerned with syntax highlighting the `css`. We will have to code a custom mode that does everything, but the `css`. I will look into making a custom mode for CodeMirror that doesn't something like this.

**Comment by Yash-Singh1** [comment]
@Gusted How does this look:
![Screenshot from 2020-11-16 16-22-04](https://user-images.githubusercontent.com/53054099/99323867-ed991400-2827-11eb-93a3-9bd5e6bf69db.png)
If you would like to preview it live:
- Go to https://pastebin.com/raw/YuZEQZ8X
- Copy the code from there
- Paste it into https://jsfiddle.net/

**Comment by Gusted** [comment]
>     * https://jsfiddle.net/

Looks pretty neat!

Some points for attention.

It's valid for CSS to have a 1 space between the selectors. e.g.
![image](https://user-images.githubusercontent.com/25481501/99459633-b3ff0080-292e-11eb-9900-d1b53ff20abd.png)
However now it's being recognized as string

Also only CSS section gets actual syntax?

For the rest :100:

**Comment by Yash-Singh1** [comment]
@Gusted Done. You can see it on https://pastebin.com/raw/YuZEQZ8X.

**Comment by Gusted** [comment]
LGTM.

1. Sites now get's weowie syntax.
2. Variables doesn't seems to get syntaxed?

**Comment by Yash-Singh1** [comment]
Hmm... I think the fix doesn't work.

**Comment by Yash-Singh1** [comment]
@Gusted Ok, it turns out that I was using a plain `string` instead of a `RegExp` 😆 
Also, I added highlighting for the selectors. To distinguish between selectors and keywords, I marked the keywords with `atom` instead. You can try it out now on the same Pastebin link.

**Comment by Gusted** [comment]
LGTM, do you want to hack around in the devtools or should I open up an PR?

**Comment by Yash-Singh1** [comment]
@Gusted Sorry for the late reply. I think the best option is for me to create a draft PR in which maintainers can edit.

**Comment by Gusted** [comment]
> @Gusted Sorry for the late reply. I think the best option is for me to create a draft PR in which maintainers can edit.

Sure things, let me know when I should take a look at the draft.

Goodluck!

Regards,
Gusted

**Comment by Yash-Singh1** [comment]
@Gusted Sure, I will tell you when I am ready with it
